File: rust/src/leetcode/Q7_reverse_integer.rs
/*
 * @lc app=leetcode id=7 lang=rust
 *
 * [7] Reverse Integer
 */

// @lc code=start
impl Solution {
    pub fn reverse(x: i32) -> i32 {
        let mut num = x;
        let mut result = 0i32;

        while num != 0 {
            if !(i32::MIN / 10..=i32::MAX / 10).contains(&result) {
                return 0;
            }

            let digit = num % 10;

            if result == i32::MAX / 10 && digit > 7 {
                return 0;
            }

            if result == i32::MIN / 10 && digit < -8 {
                return 0;
            }

            result = result * 10 + digit;
            num /= 10;
        }

        result
    }
}
// @lc code=end

File: python/Q7_reverse_integer.py
#
# @lc app=leetcode id=7 lang=python3
#
# [7] Reverse Integer
#


# @lc code=start
class Solution:
    def reverse(self, x: int) -> int:
        positive: bool = x > 0
        x = abs(x)

        result: int = 0

        while x:
            digit: int = x % 10

            if result > (0x7FFFFFFF // 10) or (
                result == (0x7FFFFFFF // 10) and digit > 7
            ):
                return 0
            if result < (-0x80000000 // 10) or (
                result == (-0x80000000 // 10) and digit > 8
            ):
                return 0

            result = (result << 3) + (result << 1) + digit

            x //= 10

        return result if positive else -result


# @lc code=end
